### ISSUE TITLE
Setting data using dynamic key doesn't update parent component template correctly

### ISSUE THREAD

**Issue by marcalexiei** [ractivejs/ractive#3234]
## Description:
While updating data in a generic component for input, parent's template and observers are not affected. However root's data has been modified.

Components involved are:
1. Root -> Form.
2. BlockGroup -> A wrapper for a group of fields. On init adds data to the object passed from `1`.
3. InputComponent -> Generic component for handling input, in this case a bulk list.

When user edit `3` data and model are correctly updated. Same happens on `2` but not in `1`.

## Versions affected:
- 0.9.x

## Platforms affected:
- Tested on Chrome 65

## Reproduction:
https://jsfiddle.net/9xdjv5t7/

```javascript
Ractive.components.InputComponent = Ractive.extend({
  template: `
  {{#each Array.from(' '.repeat(3))}}
  	<input type="checkbox" checked="{{value.includes(@index)}}" twoway="false" on-click="['set-value', @index]"/> {{@index}} <br>
  {{/each}}
  `,

  on: {
    'set-value': function(c, v) {
      const value = this.get('value');
      const index = value.indexOf(v);
      if (index > -1) {
        this.splice('value', index, 1)
      } else {
        this.push('value', v);
      }
    }
  },
});

const BlockGroup = Ractive.extend({
  template: `
  <p>BlockGroup value: {{JSON.stringify(block[key])}}</p>
  <InputComponent value="{{block[key]}}"/>
  `,
  onconfig() {
    const key = this.get('key');
    this.set(`block.${key}`, []);
  },
});

const r = window.r = new Ractive({
  el: '#main',
  components: {
    BlockGroup,
  },
  template: `
    <h1>Setting data using dynamic key doesn't update parent component template correctly</h1>
    <ol>
      <li>Check one checkbox.</li>
      <li>Only {{BlockGroup}} value is updated, root data is not updated.</li>
      <li>Click on "Preview data".</li>
      <li>Parent data is set correctly but is not updated in template.</li>
    </ol>
    <p>Root blocks is: {{JSON.stringify(blocks.foo)}}</p>
    
    <BlockGroup block="{{blocks}}" key="foo"/>
    
    <button on-click="@.set('previewData', JSON.stringify(blocks, null, 2))">Preview Root blocks</button>
    {{previewData}}
  `,
  data: {
    blocks: {
      foo: {},
    },
  },
});
```

**Comment by ghost** [comment]
Marking as bug because it looks like a bug. But I feel like I'm missing something at the moment (like a typo or something). Will dig in deeper later, unless someone gets to it first. 😁

**Comment by evs-chris** [comment]
Interestingly, this works in 0.10, which changed reference expressions to use links rather than their own free-standing proxy. Also need to change the `Root block is: {{JSON.stringify(block.foo)}}` to reference `blocks.foo`.

I'll poke at 0.9 in a bit to see if I can get he updates propagating correctly.

**Comment by dagnelies** [comment]
I also noticed that in 0.9, following works:

      const value = this.get('value');
      const index = value.indexOf(v);
      if (index > -1) {
      	   value.splice(index, 1)
      } else {
	      value.push(v);
      }
      this.set('value', value);

Instead of pushing/splicing directly

**Comment by marcalexiei** [comment]
> Also need to change the Root block is: {{JSON.stringify(block.foo)}} to reference blocks.foo.

@evs-chris Sorry, bad refactor :-/ I will update the initial comment with correct data.

@dagnelies In my app  we use `this.update`. I have tried with `this.set` but it didn't work. I'll retry tomorrow, maybe I have miss something. 

Thank you all for the quick review!

**Comment by marcalexiei** [comment]
> @dagnelies In my app we use this.update. I have tried with this.set but it didn't work. I'll retry tomorrow, maybe I have miss something.

I tried again with `this.set` and it didn't work. However in my app there is more code than in this example. Maybe there is another issue but in my code.

**Comment by evs-chris** [comment]
This is now fixed in 0.9.14. Thanks!

**Comment by marcalexiei** [comment]
Fix confirmed!
Thank you!
